### PR TITLE
Cm/state machines

### PR DESCRIPTION
**Pull request by CheezItMan**
If you have a chance to look at this I'd appreciate it.

### CODE REVIEW

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+By the end of this lesson you should be able to:
+
+- Explain what a Finite State Machine (FSM) is
+- Identify the languages a FSM can solve
+- Design a FSM to solve for a specific language pattern
+- Explain what kinds of problems a DFA cannot solve
+
+## What is a Finite State Machine
+
+A Finite State Machine is a relatively simple model of computation.  An FSA is a model of a machine which can be in exactly *one* of a limited (i.e. finite) set of states.  The machine can take in input which causes the machine to change from one state to another.  Moving from one state to another is called a **transition**.  An FSA always has one state in which it starts and one or more states which are valid end-points.  If a FSA takes in input and ends at a valid end-point, the input is considered accepted.  Otherwise the input is considered invalid and rejected.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: You use FSA and FSM throughout the lesson.  I think using FSM everywhere is probably clearer, though it would require renaming some images and if you think you're likely to say FSA or DFA out loud it may be worth defining those terms. 

```suggestion
A Finite State Machine (FSM) is a relatively simple model of computation.  An FSM is a model of a machine which can be in exactly *one* of a limited (i.e. finite) set of states.  The machine can take in input which causes the machine to change from one state to another.  Moving from one state to another is called a **transition**.  An FSM always has one state in which it starts and one or more states which are valid end-points.  If an FSM takes in input and ends at a valid end-point, the input is considered accepted.  Otherwise the input is considered invalid and rejected.  
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+- Design a FSM to solve for a specific language pattern
+- Explain what kinds of problems a DFA cannot solve
+
+## What is a Finite State Machine
+
+A Finite State Machine is a relatively simple model of computation.  An FSA is a model of a machine which can be in exactly *one* of a limited (i.e. finite) set of states.  The machine can take in input which causes the machine to change from one state to another.  Moving from one state to another is called a **transition**.  An FSA always has one state in which it starts and one or more states which are valid end-points.  If a FSA takes in input and ends at a valid end-point, the input is considered accepted.  Otherwise the input is considered invalid and rejected.  
+
+Below is a state machine that takes in input consisting of zeros (0) or ones (1), and is satisfied if the input ends with a zero.
+
+![Finite State Machine example 1](fsm-images/fsa1.png)
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: FSA -> FSM

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+Finite State Machines are used in a wide variety of applications including:
+
+- The design and verification of computer circuits
+- The design of communication protocols
+- Processing text
+- Verifying Regular Expressions
+- Developing compilers and interpreters like the Ruby interpreter
+- Modeling simple patterns of events, think of the checkout process in bEtsy
+
+FSAs can make an effective tool to help design systems which have a limited number of states, like authenticating into a website, or processing a credit card purchase.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
FSMs can make an effective tool to help design systems which have a limited number of states, like authenticating into a website, or processing a credit card purchase.
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+Knowing the rules of tennis doesn't matter, at the start the state is `Love` (score 0 for both), and if the server wins, the state becomes `15 / Love`.  If the server wins again the state becomes `30 / Love`.  Then if the opponent wins the state becomes `30 / 15`.  If the server wins twice more the state will go to `40 / 15` and then `Server Wins`.  
+
+Take a look at the state diagram, even with only 2 possible inputs at each state and a limited number of states, as the number of states increases the machine grows in complexity.  
+
+
+## A State Machine in a Table
+
+You can also represent an FSM in a table format.  The diagram and chart below illustrate the same FSM.  The left column lists the starting state and subsequent columns list the new state for the given input.  This is called a **transition table**.
+
+![Finite State Machine example 1](fsm-images/fsa1.png)
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: FSA -> FSM

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+- The design and verification of computer circuits
+- The design of communication protocols
+- Processing text
+- Verifying Regular Expressions
+- Developing compilers and interpreters like the Ruby interpreter
+- Modeling simple patterns of events, think of the checkout process in bEtsy
+
+FSAs can make an effective tool to help design systems which have a limited number of states, like authenticating into a website, or processing a credit card purchase.
+
+When you are building a system it can be a very good idea to look at the problem and see if it can be modeled as a FSM.  By doing so you can acquire a more solid understanding of the application and more easily track and identify how the application or feature can flow.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
When you are building a system it can be a very good idea to look at the problem and see if it can be modeled as an FSM.  By doing so you can acquire a more solid understanding of the application and more easily track and identify how the application or feature can flow.
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+If we have an alphabet  &Sigma; consisting of {0, 1}, and need to design an FSM to determine of the input has the pattern 1111, 1011 or 101?
+
+![Solution to example 1](fsm-images/fsm-example1.png).
+
+The trick in this case is to look at the current largest suffix of one of the string we've seen so far at each state which matches one of the patterns we are trying to match.
+
+### Example 2
+
+Now try draw a FSM which determines if 1111, 1011, 101 are contained in the input string.  You can assume the same alphabet.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
Now try draw an FSM which determines if 1111, 1011, 101 are contained in the input string.  You can assume the same alphabet.  
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+Current State | Transition Given a 1 | Transition Given a 0
+--- | --- | ---
+A | A | B
+B | A | B
+
+
+## Exercises
+
+1. Given an alphabet of {a, b} design a FSM which will determine if the input contains `aabaa`.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
1. Given an alphabet of {a, b} design an FSM which will determine if the input contains `aabaa`.
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+1. Given an alphabet of {A, C, D, G, O, T}  design a FSM which will accepts CAT or DOG alone.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
1. Given an alphabet of {A, C, D, G, O, T}  design an FSM which will accepts CAT or DOG alone.
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+1. Given an alphabet of {a, b}, design an FSM which will determine if the input has an even number of `b`s.
+
+1. Given an alphabet of {BCOTY} construct an FSM which accepts CYBOT followed by any number of characters.  
+
+1. Given an alphabet of {a, b} design a FSM which will accept an only inputs with an equal number of `a`s and `b`s.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
1. Given an alphabet of {a, b} design an FSM which will accept an only inputs with an equal number of `a`s and `b`s.
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+The study of this field is called [**Automata Theory**](https://en.wikipedia.org/wiki/Automata_theory).  It's really about what can and cannot be computed and how to represent it.  
+
+We will only touch on the 1st level of Automata theory, **Finite State Machines** (FSMs).  This is one of several ways to model computation, and a good place to start.
+
+## Learning Goals
+
+By the end of this lesson you should be able to:
+
+- Explain what a Finite State Machine (FSM) is
+- Identify the languages a FSM can solve
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
- Identify the languages an FSM can solve
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+- Design a FSM to solve for a specific language pattern
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
- Design an FSM to solve for a specific language pattern
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+8. Translate the FSM below into a table format.  You can assume things start at home.  
+  ![life Finite state machine diagram](fsm-images/life.png)
+
+9. A farm road and a highway intersect as shown below.  Detectors are placed along the farmroad to signal a vehicle is waiting to cross the highway. The traffic light controller should operate as follows. As long as no vehicle is detected on the farmroad, the lights should remain green in the highway direction. If a vehicle is detected on the farmroad, the highway lights should change from yellow to red, allowing the farmroad lights to become green. The farmroad lights stay green only as long as a vehicle is detected on the farmroad and never longer than a set interval to allow the traffic to flow along the highway. If these conditions are met, the farmroad lights change from green to yellow to red, allowing the highway lights to return to green. Even if vehicles are waiting to cross the highway, the highway should remain green for a set interval.  Design a FSM for this problem.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
9. A farm road and a highway intersect as shown below.  Detectors are placed along the farmroad to signal a vehicle is waiting to cross the highway. The traffic light controller should operate as follows. As long as no vehicle is detected on the farmroad, the lights should remain green in the highway direction. If a vehicle is detected on the farmroad, the highway lights should change from green to yellow to red, allowing the farmroad lights to become green. The farmroad lights stay green only as long as a vehicle is detected on the farmroad and never longer than a set interval to allow the traffic to flow along the highway. If these conditions are met, the farmroad lights change from green to yellow to red, allowing the highway lights to return to green. Even if vehicles are waiting to cross the highway, the highway should remain green for a set interval.  Design an FSM for this problem.
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+  Input | State symbols
+  --- | --- 
+  Farm Sensor - Timer Expired | FS - TEX
+  Farm Sensor - Timer Running | FS - TR
+  No Farm Sensor - Timer Running | NFS - TR
+  No Farm Sensor - Timer Expired | NFS - TEX
+
+**Bonus**
+
+Can you write a FSM which will determine if given the alphabet {a, b} that the input has an even number of the letter `a` or `b`?
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]:
```suggestion
Can you write an FSM which will determine if given the alphabet {a, b} that the input has an even number of the letter `a` or `b`?
```

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+1. Given an alphabet of {0, 1} design an FSM which will accept any input which does not contain `100`.
+
+1. Chris has decided to go into the smart-lock business with a keypad consisting of only 1s and 0s, along with a reset button.  He's provided the following FSM diagram to show how his lock works.
+  ![lock fsm](fsm-images/lock.png)
+
+- What is the shortest combination of keys which will open the lock?
+- Is there a longest sequence of keys which will open the lock without triggering an earlier opening?
+- Suppose Chris wanted to design a lock that required exactly 10 entries to open.  What is the minimum number of states that would be required counting the start state?
+
+8. Translate the FSM below into a table format.  You can assume things start at home.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: You switch from using `1.` for a list item to `8.` for a list item here.

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md (hunk 12 above)
Comment (kaidamasaki) [reply 04-cs-fundamentals/classroom/Finite-State-Machines.md]: Consider breaking this into smaller paragraphs.  It's a little hard to parse for humans and apparently Github just gives up on displaying a diff in the suggestion.

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+FSMs are used in a variety of applications and can be an effective way to model and understand systems which consist of a limited number of potential states.  
+
+## Further Concepts
+
+An FSM isn't a complete computational system.  Full computers can **remember** more than one particular value.  To model this mathematicains have developed additional systems which build on FSMs.
+
+### Nondeterministic Finite State Machines
+
+A nondeterministic finite state machine is a machine which can take multiple paths given a specific input.  The machine will accept input if there is *at least* one path it can follow to accept the input.  The FSMs we've been using are called Deterministic Finite State Machines as they can only take one path for each given input.  It's been shown that you can build a deterministic finite state machine that is equivalent to a nondeterministic state machine.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: This is a little bit confusing.  I think it might be worth leading with the fact that NFAs can be thought of as a more convenient way of writing FSMs since they aren't more powerful.  (Since the paragraph in "Further Concepts" sets up an expectation that NFAs are more powerful than DFAs.)

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+### Pushdown Automata
+
+A Pushdown Automata is a finite state machine that also has access to a **stack**.  Because it can access the stack to remember past inputs or states a Pushdown Automata is more powerful than a finite state machine.
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: Give an example of something a PDA can accept that a DFA can't?  Maybe the language of balanced parenthesis?

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+
+### Turing Machines
+
+[Alan Turing](https://en.wikipedia.org/wiki/Alan_Turing) is one of the father's of computer Science and developed this "complete" model of computation.  In a Turing machine the machine has an infinitely long "tape" of cells.  It can move to individual cells, read and write the value at each cell.  With this the machine has infinite storage capcacity.  In real life no machine has infinite storage, but modern computer systems usually have much more than is needed, and a Turing Machine has been shown to be able to compute an answer to any problem a computer can solve (speed is another issue).  Thus the term **Turing Complete** is used to indicate a programming language that can accomplish all tasks that a computer can accomplish, if you ignore the limitations of memory.  
+
+Turing machines were used in mathematical proofs to illustrate a class of problems which are not computable.  In other words, you cannot write a computer program to solve them in all cases.  A classic example is that you cannot write a program which will read in another program and predict with perfect accuracy if that program will ever end (no infinite loops).
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: IIRC it's not even possible to probabilistically predict whether or not a machine will halt correctly.  (I asked a professor about this once but it was a while ago so perhaps I'm mis-remembering or the state of research has changed.)

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+## Resources
+
+- [Free Stamford Course on Finite Automata]()
```
Comment (kaidamasaki) [inline 04-cs-fundamentals/classroom/Finite-State-Machines.md]: Missing link.

File: 04-cs-fundamentals/classroom/Finite-State-Machines.md (hunk 17 above)
Comment (CheezItMan) [reply 04-cs-fundamentals/classroom/Finite-State-Machines.md]: You can't predict if it will halt correctly or even if will ever  halt actually.  I had to do the proof on this for my final once.
